csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -441,295 +405,463 @@ async Task StartedAsync(IServiceObserverPrx observer, string[] services)
                     }
-                    throw;
+                    catch (Exception ex)
+                    {
+                        serviceManager._logger.Warning(
+                            $"IceBox.ServiceManager: exception while starting service `{Name}':\n{ex}");
+                        throw new ServiceException($"service {Name} failed to start", ex);
+                    }
+
+                    // We send the notification (without waiting for the response) just before the task completes
+                    // successfully, to ensure proper ordering of notifications. While the task is still running, it
+                    // cannot get replaced by a new task.
+                    serviceManager.ServiceStarted(Name);
                 }
             }
-        }
 
-        private void StopAll()
-        {
-            lock (_mutex)
+            /// <summary>Stops this service.</summary>
+            /// <param name="serviceManager">The service manager.</param>
+            /// <param name="notifyObservers">When true, the new stop task notifies the observers when the service is
+            /// stopped; when false, the new stop task does not notify observers.</param>
+            /// <return>A task that completes successfully when the service is fully stopped.</return>
+            internal async Task StopAsync(ServiceManager serviceManager, bool notifyObservers)
             {
-                // First wait for any active startService/stopService calls to complete.
-                while (_pendingStatusChanges)
-                {
-                    Monitor.Wait(_mutex);
-                }
+                Task? stopTask = null;
 
-                // For each service, we call stop on the service and flush its database environment to
-                // the disk. Services are stopped in the reverse order of the order they were started.
-                _services.Reverse();
-                var stoppedServices = new List<string>();
-                foreach (ServiceInfo info in _services)
+                do
                 {
-                    if (info.Status == ServiceStatus.Started)
+                    // We may need to wait until the current start task completes.
+                    Task? startTask = null;
+
+                    lock (serviceManager._mutex)
                     {
-                        try
+                        if (IsStarted)
                         {
-                            Debug.Assert(info.Service != null);
-                            info.Service.Stop();
-                            stoppedServices.Add(info.Name);
+                            // If it's currently being stopped, we piggy-back on this stop, else we create a new stop
+                            // task.
+                            _pendingTask ??= PerformStopAsync();
+                            stopTask = _pendingTask;
                         }
-                        catch (Exception ex)
+                        else
```

Comment:
You can use a `else if\else` block instead of nesting `if\else` in the `else`

Reply:
I prefer to keep the symmetry with `StartAsync` where `else if` does not work as well as it would require an ugly `if (!IsStarted)`.